### PR TITLE
BUG:  DataFrame.update bool dtype being converted to object #55509

### PR DESCRIPTION
**Pull request by aureliobarbosa**
- [x] Fixes #55509
- [x] Tests added and passed.
- [x] All code checks passed.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by aureliobarbosa** [comment]
This is my first PR. For some reason that I still can't understand the commit that was immediately before the one I submitted got resubmitted. I imagine it can be something usual and easy to fix, but let me know if it is necessary that I  fix something (either on GitHub or on my local repository).

**Comment by aureliobarbosa** [comment]
@rhshadrach

In short. To capture this bug I analyzed the "Unit Tests / Copy-on-Write 3.9 (pull_request)" and captured the single test giving the warning. After studying it it was possible to reproduce this test on my machine using the same pytest called by you on the GitHub CI with a proper environment setup for testing "copy-on-write". For the latter I tested using both the environment variable and `pandas.set_option()`. 

In the pull request above the code is failing on `FutureWarning`, and I believe that my code is giving the proper results (which were check through local builds), but that the test is outdated (it refer to #3217, which is from before 2013). 

Below I put the actual test that I think is wrong,  and the new proposed test (for which my new code on update passes):

File: `pandas/tests/frame/methods/test_update.py`
(Possibly) wrong test:
```python 
    def test_update_with_different_dtype(self, using_copy_on_write):
        # GH#3217
        df = DataFrame({"a": [1, 3], "b": [np.nan, 2]})
        df["c"] = np.nan
        if using_copy_on_write:
            df.update({"c": Series(["foo"], index=[0])})
        else:
            with tm.assert_produces_warning(FutureWarning, match="incompatible dtype"):
                df["c"].update(Series(["foo"], index=[0]))

        expected = DataFrame({"a": [1, 3], "b": [np.nan, 2], "c": ["foo", np.nan]})
        tm.assert_frame_equal(df, expected)
```

Curren proposed test:
```python 
    def test_update_with_different_dtype(self, using_copy_on_write):
        # GH#3217 #55509
        df = DataFrame({"a": [1, 3], "b": [np.nan, 2]})
        df["c"] = np.nan
        with tm.assert_produces_warning(FutureWarning, match="incompatible dtype"):
            if using_copy_on_write:
                df.update({"c": Series(["foo"], index=[0])})
            else:
                df["c"].update(Series(["foo"], index=[0]))

        expected = DataFrame({"a": [1, 3], "b": [np.nan, 2], "c": ["foo", np.nan]})
        tm.assert_frame_equal(df, expected)
```


Given this information I am not sure how to proceed:
1) Open an issue to correct the test. Wait for review.
2) Make a new pull request fixing test and the new code (separating the commits for history checks).
3) Close this pull request and wait for further discussion among core developers. 

Regards

File: pandas/core/frame.py
Comment (rhshadrach) [inline pandas/core/frame.py:8860]: For a monotonic index, I'm seeing a perf increase

```
size = 100_000
df1 = pd.DataFrame({'a': np.random.randint(0, 100, size)}, index=range(1, size+1))
df2 = pd.DataFrame({'a': np.random.randint(0, 100, 3)}, index=[10, 12, 11])

%timeit df1.update(df2)
# 1.33 ms ± 16.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- main
# 372 µs ± 663 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)    <-- PR
```

But a perf regression for nonmonotonic:

```
size = 100_000
index = list(range(1, size+1))
index[1] = index[0]
df1 = pd.DataFrame({'a': np.random.randint(0, 100, size)}, index=index)
df2 = pd.DataFrame({'a': np.random.randint(0, 100, 3)}, index=[10, 12, 11])
%timeit df1.update(df2)
# 1.1 ms ± 3.72 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- main
# 2.52 ms ± 7.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)    <-- PR
```

And a perf regression when the caller's index is small and the argument's index is large:

```
size = 100_000
df1 = pd.DataFrame({'a': np.random.randint(0, 100, 3)}, index=[10, 11, 12])
df2 = pd.DataFrame({'a': np.random.randint(0, 100, size)})
%timeit df1.update(df2)
# 186 µs ± 997 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)  <-- main
# 329 µs ± 595 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)   <-- PR
```

When there are duplicates and the index of the caller and argument are the same, both main and this PR succeed. When they are not the same, both fail but with different error messages

```
size = 10
df1 = pd.DataFrame({'a': np.random.randint(0, 100, size)})
df2 = pd.DataFrame({'a': [-1, -2, -3]}, index=[4, 4, 6])
df1.update(df2)
# ValueError: cannot reindex on an axis with duplicate labels  <-- main
# ValueError: operands could not be broadcast together with shapes (3,) (2,) (3,)  <-- PR
```

Also, when there is no overlap (the intersection is empty), this PR succeeds whereas main fails with the same error message.

Comment (rhshadrach) [reply pandas/core/frame.py:8860]: A bit mixed when it comes to performance, but I'd guess sorted indices and having a large caller with a small updater is the common cause, and this is where it's an improvement.

My recommendation would be to detect duplicate indices (in either the caller or the argument) and raise an informative error message, even when we would otherwise succeed (when there is no overlap).

File: pandas/core/generic.py
Comment (rhshadrach) [inline pandas/core/generic.py:8796]: Unrelated?

File: pandas/tests/frame/methods/test_update.py
Comment (rhshadrach) [inline pandas/tests/frame/methods/test_update.py:181]: Can you add a test for when the index has duplicate values

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looking good! I think we should add the test from the issue (https://github.com/pandas-dev/pandas/issues/55509#issuecomment-1767324695).

File: doc/source/whatsnew/v2.1.2.rst
Comment (rhshadrach) [inline doc/source/whatsnew/v2.1.2.rst:30]: This isn't a regression, so would go into 2.2.0.

File: pandas/core/frame.py
Comment (rhshadrach) [reply pandas/core/frame.py:8860]: cc @mroeschke - do you agree?

**Comment by aureliobarbosa** [comment]
> Looking good! I think we should add the test from the issue ([#55509 (comment)](https://github.com/pandas-dev/pandas/issues/55509#issuecomment-1767324695)).

Thanks for the return. I will add a test similar to the one presented in the discussion.

**Comment by aureliobarbosa** [comment]
Organizing ideas:

Tests: 
-    [ ] Add tests for when the index has duplicate values.
-    [ ] Add the longer test checking various types, from issue #55509.

Investigate: 
-    [ ] implementation of `DataFrame.clip()`, for ideas.
-    [ ] performance regressions nonmonotonic case.
-    [ ] performance regressions small dataframe with larger dataframe on argument.

Error messages:
-  [ ] duplicate indexes.
-  [ ] no overlap.

Docs:
-  [ ] Change whatsnew.rst, version 2.2

Questions:

- Was copy-on-write used on the tests above?
- Should strings be included in the longer test?
- What should I do about `test_update_with_different_dtype()`, discussed above?

File: pandas/core/frame.py
Comment (mroeschke) [reply pandas/core/frame.py:8860]: Yeah I think that makes sense to me. The error message for duplicates in this PR is not great compared to the error message on main.

File: pandas/core/frame.py
Comment (mroeschke) [inline pandas/core/frame.py:8863]: Could we avoid chained indexing here?

File: pandas/core/frame.py
Comment (rhshadrach) [inline pandas/core/frame.py:8874]: Can you add a message as to why order matters

File: pandas/core/frame.py
Comment (rhshadrach) [inline pandas/core/frame.py:8879]: Isn't this not backwards compatible? I think this case does not raise on main (in general), is that right?

File: pandas/core/frame.py
Comment (rhshadrach) [inline pandas/core/frame.py:8881]: [I've recommended](https://github.com/pandas-dev/pandas/pull/55634#discussion_r1375265143) to raise in this case. Do you disagree?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Apologies for the delay @aureliobarbosa - this fell of my radar.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by rhshadrach** [comment]
@aureliobarbosa - okay if I finish this up for 3.0?

**Comment by aureliobarbosa** [comment]
@rhshadrach Thanks for bringing it back. Would you mind if I try to solve it until Feb 20th?

**Comment by rhshadrach** [comment]
Sure!

**Comment by aureliobarbosa** [comment]
@rhshadrach Tests are passing on my setup but the commit history of this PR got very noised. I would like to move it to a new PR, making reference to this. Is it OK?

**Comment by rhshadrach** [comment]
@aureliobarbosa - I think in this case that's fine. For longer PRs (e.g. something like 300 line diffs) it can create a lot of rework for reviewers, but that's not the case here.

**Comment by aureliobarbosa** [comment]
@rhshadrach Done here. 

The intention of the  commit below was just a merge but I incorporate the main code fix on it (single line), by mistake.
[Merge remote-tracking branch 'upstream/main' into fix_gh55509](https://github.com/pandas-dev/pandas/pull/55634/commits/ce1b7a764d00ba6b88c84e9ed86e1278e64dbf33)

Also, I don't understand what is wrong with the doctest. 

I am available for any question or fix.

File: pandas/core/frame.py
Comment (rhshadrach) [reply pandas/core/frame.py:8781]: I don't understand my question here - looks like I might have read the code as being `if not other.index.is_unique`.

**Comment by rhshadrach** [comment]
@aureliobarbosa - thanks for the update. There are 2 questions/requests above that should be addressed.

> Also, I don't understand what is wrong with the doctest.

The doctest is failing on `DataFrame.to_xarray` and this method does use `DataFrame.update`. It looks like this needs to be addressed.

**Comment by aureliobarbosa** [comment]
@rhshadrach Thanks for the feedback. I will try to address this issue this weekend.

**Comment by mroeschke** [comment]
Closing in favor of https://github.com/pandas-dev/pandas/pull/57637

**Pull request closed without merging**
